Source/SwiftLintBuiltInRules/Rules/Metrics/NestingRule.swift

```diff
@@ -12,137 +13,161 @@ struct NestingRule: Rule {
-        .union([.enumcase, .enumelement])
-        .map(SwiftStructureKind.declaration)
-
-    private struct ValidationArgs {
-        var typeLevel: Int = -1
-        var functionLevel: Int = -1
-        var previousKind: SwiftStructureKind?
-        var violations: [StyleViolation] = []
+private struct ValidationData {
+    private(set) var typeLevel: Int = -1
+    private(set) var functionLevel: Int = -1
+    private(set) var functionOrNotStack = Stack<Bool>()
 
-        func with(previousKind: SwiftStructureKind?) -> ValidationArgs {
-            var args = self
-            args.previousKind = previousKind
-            return args
-        }
+    mutating func push(_ isFunction: Bool) {
+        functionOrNotStack.push(isFunction)
+        updateLevel(with: 1, isFunction: isFunction)
     }
 
-    func validate(file: SwiftLintFile) -> [StyleViolation] {
-        return validate(file: file, substructure: file.structureDictionary.substructure, args: ValidationArgs())
+    mutating func pop() {
+        updateLevel(with: -1, isFunction: functionOrNotStack.pop()!)
     }
 
-    private func validate(file: SwiftLintFile, substructure: [SourceKittenDictionary],
-                          args: ValidationArgs) -> [StyleViolation] {
-        return args.violations + substructure.flatMap { dictionary -> [StyleViolation] in
-            guard let kindString = dictionary.kind, let structureKind = SwiftStructureKind(kindString) else {
-                return validate(file: file, substructure: dictionary.substructure, args: args.with(previousKind: nil))
-            }
-            guard !omittedStructureKinds.contains(structureKind) else {
-                return args.violations
-            }
-            switch structureKind {
-            case let .declaration(declarationKind):
-                if configuration.ignoreTypealiasesAndAssociatedtypes,
-                   declarationKind == .associatedtype || declarationKind == .typealias {
-                    return args.violations
-                }
-                return validate(file: file, structureKind: structureKind,
-                                declarationKind: declarationKind, dictionary: dictionary, args: args)
-            case .expression, .statement:
-                guard configuration.checkNestingInClosuresAndStatements else {
-                    return args.violations
-                }
-                return validate(file: file, substructure: dictionary.substructure,
-                                args: args.with(previousKind: structureKind))
-            }
+    private mutating func updateLevel(with value: Int, isFunction: Bool) {
+        if isFunction {
+            functionLevel += value
+        } else {
+            typeLevel += value
         }
     }
+}
+
+private extension NestingRule {
+    final class Visitor: ViolationsSyntaxVisitor<ConfigurationType> {
+        override var skippableDeclarations: [any DeclSyntaxProtocol.Type] { [ProtocolDeclSyntax.self] }
```

Comment:
This is contra productive later when nested protocols are allowed. Perhaps you can add a test for this case already.